### PR TITLE
[Core] Add pGetGeometryPart BrepCurveOnSurface

### PR DESCRIPTION
**Pull request by tteschemacher**
**Description**
Adds pGetGeometryPart to the BrepCurveOnSurface and corresponding NurbsCurveOnSurface

Helps to obtain the access to the underlying embedded parameter curve.

### CODE REVIEW

File: kratos/geometries/brep_curve_on_surface.h [hunk 1]
```diff
@@ -212,6 +216,59 @@ class BrepCurveOnSurface
...
+    *        which is corresponding to the index.
+    *        Possible indices are:
+    *        SURFACE_INDEX, EMBEDDED_CURVE_INDEX or CURVE_ON_SURFACE_INDEX.
+    * @param Index: SURFACE_INDEX, EMBEDDED_CURVE_INDEX or CURVE_ON_SURFACE_INDEX.
+    * @return pointer of geometry, corresponding to the index.
+    */
+    GeometryPointer pGetGeometryPart(const IndexType Index) override
+    {
+        const auto& const_this = *this;
+        return std::const_pointer_cast<GeometryType>(
```
Comment (philbucher) [inline kratos/geometries/brep_curve_on_surface.h:234]: what is this doing? :sweat_smile:  :sweat_smile:  :sweat_smile:

File: kratos/geometries/nurbs_curve_on_surface_geometry.h [hunk 2]
```diff
@@ -155,6 +160,56 @@ class NurbsCurveOnSurfaceGeometry : public Geometry<typename TSurfaceContainerPo
...
+    *        which is corresponding to the trim index.
+    *        Possible indices are:
+    *        SURFACE_INDEX or EMBEDDED_CURVE_INDEX.
+    * @param Index: SURFACE_INDEX or EMBEDDED_CURVE_INDEX.
+    * @return pointer of geometry, corresponding to the index.
+    */
+    GeometryPointer pGetGeometryPart(const IndexType Index) override
+    {
+        const auto& const_this = *this;
+        return std::const_pointer_cast<GeometryType>(
```
Comment (philbucher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h:178]: same

File: kratos/geometries/nurbs_curve_on_surface_geometry.h [hunk 3]
```diff
@@ -155,6 +160,56 @@ class NurbsCurveOnSurfaceGeometry : public Geometry<typename TSurfaceContainerPo
...
+    *        SURFACE_INDEX.
+    * @param Index of the geometry part.
+    * @return true if SURFACE_INDEX.
+    */
+    bool HasGeometryPart(const IndexType Index) const override
+    {
+        if (Index == SURFACE_INDEX)
+            return true;
+
+        return false;
```
Comment (philbucher) [inline kratos/geometries/nurbs_curve_on_surface_geometry.h]:
```suggestion
        return Index == SURFACE_INDEX;
```

File: kratos/geometries/brep_curve_on_surface.h [hunk 4]
```diff
@@ -212,6 +216,59 @@ class BrepCurveOnSurface
...
+    *        SURFACE_INDEX or CURVE_ON_SURFACE_INDEX.
+    * @param Index of the geometry part.
+    * @return true if SURFACE_INDEX or CURVE_ON_SURFACE_INDEX.
+    */
+    bool HasGeometryPart(const IndexType Index) const override
+    {
+        if (Index == SURFACE_INDEX || Index == CURVE_ON_SURFACE_INDEX)
+            return true;
+
+        return false;
```
Comment (philbucher) [inline kratos/geometries/brep_curve_on_surface.h]:
```suggestion
        return (Index == SURFACE_INDEX || Index == CURVE_ON_SURFACE_INDEX);
```

File: kratos/geometries/nurbs_curve_on_surface_geometry.h (hunk 2 above)
Comment (tteschemacher) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:178]: Const casting the const. It does avoid coffee duplication.

Comment (philbucher) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:178]: ahh now I understand what this does
you don't want to implement the function twice, once for with and once for without `const`

Usually we duplicate the code but this is already done in other places so fine for me
If we want to change it at some point would be easy

**Review by philbucher — APPROVED** [review]

File: kratos/geometries/nurbs_curve_on_surface_geometry.h (hunk 2 above)
Comment (tteschemacher) [reply kratos/geometries/nurbs_curve_on_surface_geometry.h:178]: Ups too much coffee und too less sleep ;)

Yes, it's not really beautiful. But it does it's job...

**Pull request merged**